### ISSUE TITLE
Can not refine types generically

### ISSUE THREAD

**Issue by adrian-budau** [TSOptics/optics#18]
Sorry for the weird title, did not know how to properly phrase this.

The big issues I found trying to use the library is with writing generic extra utilities 

Problem 1:
```ts
function asRead<T>(optic: Optic<T>): ReadOptic<T> {
  return optic;
}
```
Somehow Optic<number> is assignable to ReadOptic<number>, and any other example where I instantiate T, but generically does not work, I have not found a workaround for this.

Problem 2 (I found a solution)
```ts
function complexUseOptic<T>(optic: Optic<T>) {
  const [value, ..._rest] : [T, ...unknown[]] = useOptic(optic);
}
```
Problem 2 seems to be solved by just writing a much longer type for useOptic, I've managed to reduce the problem to the fact that `[GetOpticFocus<U>] extends [infer TFocus] ? TFocus : never' is not assignable to type 'GetOpticFocus<U>'`.

This could be solved by just removing the infer from useOptic and using the explicit `GetOpticFocus` and `GetOpticScope`, but there might be a better solution.

**Comment by VincentQuillien** [comment]
Hello @adrian-budau, thanks for the feedback !

I've just published v0.5.0 where `ReadOptic` and `AsyncOptic` are replaced by modifiers that we apply to the second type parameter of the optic: https://github.com/TSOptics/optics/pull/20.

Now that we have only a single optic type, writing function with generic focused type works like you expect.

```ts
function asRead<T>(optic: Optic<T>): Optic<T, readOnly> {
  return optic;
}
```

`useOptic` should also now properly preserve the type param.
Tell me if it works for you.

**Comment by adrian-budau** [comment]
Thank you, it works as expected!
